src/ucp/tag/rndv.c

```diff
@@ -5,19 +5,316 @@
+    ucp_request_t *req     = get_req->send.rndv_get.rreq;
+    uintptr_t remote_req   = get_req->send.rndv_get.remote_request;
+
+    ucs_debug("rndv get completed");
+
+    ucp_request_complete_recv(req, UCS_OK, &req->recv.info);
+    ucp_request_send_buffer_dereg(get_req, ucp_ep_get_rndv_data_lane(get_req->send.ep));
+
+    get_req->send.lane         = ucp_ep_get_am_lane(get_req->send.ep);
+    get_req->send.uct.func     = ucp_proto_progress_am_bcopy_single;
+    get_req->send.proto.am_id  = UCP_AM_ID_RNDV_ATS;
+    get_req->send.proto.status = UCS_OK;
+    get_req->send.proto.remote_request = remote_req;
+
+    ucs_debug("send rndv ats");
+    ucp_request_start_send(get_req);
+}
+
+void ucp_rndv_matched(ucp_worker_h worker, ucp_request_t *req,
+                      ucp_rndv_rts_hdr_t *rndv_rts_hdr)
+{
+    uct_rkey_bundle_t rkey_ob;
+    ucp_request_t *get_req;
+    size_t recv_size;
+
+    UCS_ASYNC_BLOCK(&worker->async);
+
+    ucs_trace_req("ucp_rndv_matched");
+
+    req->recv.info.sender_tag = rndv_rts_hdr->super.tag;
+    req->recv.info.length     = rndv_rts_hdr->size;
+
+    get_req = ucp_worker_allocate_reply(worker, rndv_rts_hdr->req.sender_uuid);
+    get_req->send.uct.func     = ucp_proto_progress_rndv_get;
+    get_req->send.buffer       = req->recv.buffer;
+
+    ucs_assert_always(req->recv.count != 0);
+    ucs_assertv_always((req->recv.datatype & UCP_DATATYPE_CLASS_MASK) == UCP_DATATYPE_CONTIG,
+                       "dtype=0x%lx", req->recv.datatype);
+
+    uct_rkey_unpack(rndv_rts_hdr + 1, &rkey_ob);
+    get_req->send.rndv_get.remote_request = rndv_rts_hdr->req.reqptr;
+    get_req->send.rndv_get.remote_address = rndv_rts_hdr->address;
+    get_req->send.rndv_get.rkey = rkey_ob.rkey;
+    get_req->send.rndv_get.rreq = req;
+
+    recv_size = ucp_contig_dt_length(req->recv.datatype, req->recv.count);
+    if (ucs_unlikely(recv_size < rndv_rts_hdr->size)) {
+        ucs_debug("message truncated: recv_length %zu rts_send_size %zu",
+                  recv_size, rndv_rts_hdr->size);
+        get_req->send.uct.func  = ucp_rndv_truncated;
+        get_req->send.lane      = ucp_ep_get_am_lane(get_req->send.ep);
+        goto out;
+    }
+    get_req->send.length         = rndv_rts_hdr->size;
+    get_req->send.uct_comp.func  = ucp_rndv_get_completion;
+    get_req->send.state.offset   = 0;
+    get_req->send.lane           = ucp_ep_get_rndv_data_lane(get_req->send.ep);
+    get_req->send.state.dt.contig.memh = UCT_INVALID_MEM_HANDLE;
+
+    /* if the receive side is not connected yet then the RTS was received on a stub ep */
+    if (ucp_ep_is_stub(get_req->send.ep)) {
+        ucs_debug("received rts on a stub ep, ep=%p, rndv_lane=%d",
+                  get_req->send.ep, ucp_ep_get_rndv_data_lane(get_req->send.ep));
+    }
+
+out:
+    /* Release the remote key since its content was copied to the get request
+     * for the future get operation */
+    uct_rkey_release(&rkey_ob);
```

Comment:
cannot release the rkey here because the get_req may fail (not happens with IB, but will happen in other transports)